### DISCUSSION TITLE
Psycopg 3 cast auto binding

### DISCUSSION THREAD

**Discussion by dukkee** [sqlalchemy/sqlalchemy#10839] [Usage Questions]
Hello, I would be happy if somebody could explain to me why the cast auto-binding was implemented for psycopg 3 (psycopg 2 doesn't have it). I mean the types addition to the fields in the `WHERE` clause e.g.
I don't need it and it forces me to write extra code to cover `string -> int` conversion in some places.
I would also appreciate it if somebody could explain if it can be turned off.

```sql
SELECT * from my_table where my_field = %(value_1)s::VARCHAR
```

**Comment by zzzeek** [comment]
im not sure what "it forces me to write extra code to cover string->int" means, in SQLAlchemy you always send the exact Python type for any SQL statement so it appears you might be relying on some implicit behavior in psycopg2, I'd need to see explicitly what it is you're trying to do.

The cast is needed because psycopg uses prepared statements and there are many cases where postgresql refuses to accept simple integers and other types without a type cast.  Try this statement with psycopg:

```py
import psycopg

conn = psycopg.connect(user="scott", password="tiger", host="localhost", dbname="test")
cursor = conn.cursor()

cursor.execute("""

SELECT x
FROM generate_series(%(generate_series_1)s, %(generate_series_2)s) AS x

""", {"generate_series_1": 1, "generate_series_2": 10})
```

output:

```
psycopg.errors.AmbiguousFunction: function generate_series(smallint, smallint) is not unique
LINE 4: FROM generate_series($1, $2) AS x
             ^
HINT:  Could not choose a best candidate function. You might need to add explicit type casts.

```

see the problem?    that statement works as is in psycopg2.

I dont really see what the issue with SQLAlchemy's generated statements having type casts would be.  If you are trying to send strings in and have them be automatically interpreted as ints, that's wrong.

**Reply by dukkee** [reply]
> If you are trying to send strings in and have them be automatically interpreted as ints, that's wrong.

Hello @zzzeek Thank you for your answer. 

Yeah, it's that case. In one of the old projects, which I'm trying to migrate to the Sqlalchemy 2 + psycopg 3, there is no separate layer to convert some ids from the strings to ints and somewhere it's passed as is. It worked in psycopg 2 and it works fine now with version 3 if don't mention the cast. 

I know that the values should be converted to the correct type, but it's a question of the actual tasks priorities.

Okay, so this is done to decrease the amount of cases when the developer needs to write additional type casts, right?

**Reply by CaselIT** [reply]
> Okay, so this is done to decrease the amount of cases when the developer needs to write additional type casts, right?

Yes

Note that sqlalchemy 2 supports psycopg2 so you can keep using that

**Reply by zzzeek** [reply]
you should also be aware that psycopg2 rendered type casts on its own, so there's not much difference in the end result here, just we might be doing different casts that you don't prefer.

given my program above, if we look at what is actually sent by **psycopg2** to PostgreSQL looking at server logs, it is auto-casting:



```
2024-01-07 19:06:45.446 EST [30371] LOG:  execute <unnamed>: SELECT x 
	FROM generate_series($1::INTEGER, $2::INTEGER) AS x
2024-01-07 19:06:45.446 EST [30371] DETAIL:  parameters: $1 = '1', $2 = '2'

```

whereas newer psycopg does this:

```
2024-01-07 19:08:24.295 EST [30473] LOG:  execute <unnamed>: BEGIN
2024-01-07 19:08:24.295 EST [30473] ERROR:  function generate_series(smallint, smallint) is not unique at character 17
2024-01-07 19:08:24.295 EST [30473] HINT:  Could not choose a best candidate function. You might need to add explicit type casts.
2024-01-07 19:08:24.295 EST [30473] STATEMENT:  
	
	SELECT x
	FROM generate_series($1, $2) AS x

```

hence SQLAlchemy has to take over the job of casting.

still not sure what case "worked" with psycopg2 that doesnt with psycopg, specifics would help.  psycogp2 would be casting based on what's given, not what's wanted.

**Reply by dukkee** [reply]
@zzzeek @CaselIT thank you a lot for your comments. As I mentioned in the start message, the case is quite trivial, there is a model with an integer field and it's needed to be filtered by this field. Something like this one:
```python
class MyModel(Model):
    __tablename__ = "my_model"

    id = Column(String, primary_key=True, default=get_uuid)
    my_field = Column(BigInteger)


print(MyModel.query.filter(MyModel.my_field == '1'))
# SELECT * from my_model WHERE my_model.id = %(my_field_1)s::VARCHAR
```

As now the cast that is added to the select is based on the passed value type and not on the column type, it starts to conflict. If it was `::INTEGER`, it would be used by psycopg3 without error.

**Reply by isolovey** [reply]
> If you are trying to send strings in and have them be automatically interpreted as ints, that's wrong.

>still not sure what case "worked" with psycopg2 that doesnt with psycopg, specifics would help. psycogp2 would be casting based on what's given, not what's wanted.

@zzzeek This is done by [Flask-Admin](https://github.com/flask-admin/flask-admin), (e.g. not casting an id column value as integer), and it works fine with `psycopg2`. While I understand that this is likely an issue that should be addressed in Flask-Admin, that project doesn't seem to be very actively maintained.

**Reply by zzzeek** [reply]
psycopg2 is doing the casts for you behind the scenes.     I would take a look at your [server side SQL logs](https://www.postgresql.org/docs/current/runtime-config-logging.html) to see what psycopg2 sends vs. what SQLAlchemy's casts with psycopg3 look like, maybe there is some insight there.
